lib/ansible/modules/cloud/azure/azure_rm_postgresqldatabase.py

```diff
@@ -0,0 +1,276 @@
+
+    def __init__(self):
+        self.module_arg_spec = dict(
+            resource_group=dict(
+                type='str',
+                required=True
+            ),
+            server_name=dict(
+                type='str',
+                required=True
+            ),
+            name=dict(
+                type='str',
+                required=True
+            ),
+            charset=dict(
+                type='str'
+            ),
+            collation=dict(
+                type='str'
+            ),
+            state=dict(
+                type='str',
+                default='present',
+                choices=['present', 'absent']
+            )
+        )
+
+        self.resource_group = None
+        self.server_name = None
+        self.name = None
+        self.parameters = dict()
+
+        self.results = dict(changed=False)
+        self.mgmt_client = None
+        self.state = None
+        self.to_do = Actions.NoAction
+
+        super(AzureRMDatabases, self).__init__(derived_arg_spec=self.module_arg_spec,
+                                               supports_check_mode=True,
+                                               supports_tags=False)
+
+    def exec_module(self, **kwargs):
+        """Main module execution method"""
+
+        for key in list(self.module_arg_spec.keys()):
+            if hasattr(self, key):
+                setattr(self, key, kwargs[key])
+            elif kwargs[key] is not None:
+                if key == "charset":
+                    self.parameters["charset"] = kwargs[key]
+                elif key == "collation":
+                    self.parameters["collation"] = kwargs[key]
+
+        old_response = None
+        response = None
+
+        self.mgmt_client = self.get_mgmt_svc_client(PostgreSQLManagementClient,
+                                                    base_url=self._cloud_environment.endpoints.resource_manager)
+
+        resource_group = self.get_resource_group(self.resource_group)
+
+        old_response = self.get_postgresqldatabase()
+
+        if not old_response:
+            self.log("PostgreSQL Database instance doesn't exist")
+            if self.state == 'absent':
+                self.log("Old instance didn't exist")
+            else:
+                self.to_do = Actions.Create
+        else:
+            self.log("PostgreSQL Database instance already exists")
+            if self.state == 'absent':
+                self.to_do = Actions.Delete
+            elif self.state == 'present':
+                self.log("Need to check if PostgreSQL Database instance has to be deleted or may be updated")
+                self.to_do = Actions.Update
+                self.delete_postgresqldatabase()
+
+        if (self.to_do == Actions.Create) or (self.to_do == Actions.Update):
+            self.log("Need to Create / Update the PostgreSQL Database instance")
+
+            if self.check_mode:
```

Comment:
This behaviour is really not Ansible like, it seems to indicate if a create or "update" will report a change which is ok but the trouble is that we aren't checking if an update is actually required. So a check mode run will always report a change if the database exists. We should be putting a check above this to verify that if the database exists, then we check the details like `charset`, `collation` to see if they match our parameters. If they do then no update is required, if they don't then we would report a change.

This module makes things a bit difficult as I'm not sure you can change the collation and charset after it has been created so we would need to error out or display a warning saying we couldn't update it.

Reply:
ok, i just added check for charset and collation.

Regarding update, at the moment I am working with the team responsible for these services to figure out whether / when this issue can be resolved, so i more treat it as a temporary issue right now. Perhaps we should add appropriate note.

Reply:
This comment was more around the scenario when

* When the database exists
* State is `present`
* Running in check mode

According to the logic a change will always be reported in this scenario which may not necessarily be true.